examples/advanced_operations/add_gmail_ad.py

```diff
@@ -0,0 +1,133 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This example adds a gmail ad to a given ad group.
+
+The ad group's campaign needs to have an AdvertisingChannelSubtype of
+DISPLAY_GMAIL_AD. To get ad groups, run basic_operations/get_ad_groups.py
+"""
+
+from __future__ import absolute_import
+
+import argparse
+import six
+import sys
+from uuid import uuid4
+import requests
+
+from google.ads.google_ads.client import GoogleAdsClient
+from google.ads.google_ads.errors import GoogleAdsException
+
+LOGO_URL = 'https://goo.gl/mtt54n'
+MARKETING_IMG_URL = 'http://goo.gl/3b9Wfh'
+
+
+def main(client, customer_id, ad_group_id):
+    logo_img_bytes, logo_img_content_type = get_image(LOGO_URL)
+    marketing_img_bytes, marketing_img_content_type = get_image(
+        MARKETING_IMG_URL)
+
+    if logo_img_content_type != 'image/png':
+        raise ValueError('Logo image has invalid content-type.')
+
+    if marketing_img_content_type != 'image/jpeg':
+        raise ValueError('Marketing image has invalid content-type.')
+
+    media_file_logo_op = client.get_type('MediaFileOperation')
+    media_file_logo = media_file_logo_op.create
+    media_file_logo.type = client.get_type('MediaTypeEnum').IMAGE
+    media_file_logo.image.data.value = logo_img_bytes
+    media_file_logo.mime_type = client.get_type('MimeTypeEnum').IMAGE_PNG
+
+    media_file_marketing_op = client.get_type('MediaFileOperation')
+    media_file_marketing = media_file_marketing_op.create
+    media_file_marketing.type = client.get_type('MediaTypeEnum').IMAGE
+    media_file_marketing.image.data.value = marketing_img_bytes
+    media_file_marketing.mime_type = client.get_type('MimeTypeEnum').IMAGE_JPEG
+
+    media_file_service = client.get_service('MediaFileService')
+    image_response = media_file_service.mutate_media_files(
+        customer_id, [media_file_logo_op, media_file_marketing_op])
+
+    image_resource_names = list(map(lambda response: response.resource_name,
+                                    image_response.results))
+
+    ad_group_ad_service = client.get_service('AdGroupAdService')
+    ad_group_ad_op = client.get_type('AdGroupAdOperation')
+    ad_group_ad = ad_group_ad_op.create
+    gmail_ad = ad_group_ad.ad.gmail_ad
+    gmail_ad.teaser.headline.value = 'Dream'
+    gmail_ad.teaser.description.value = 'Create your own adventure.'
+    gmail_ad.teaser.business_name.value = 'Interplanetary Ships'
+    gmail_ad.teaser.logo_image.value = image_resource_names[0]
+    gmail_ad.marketing_image.value = image_resource_names[1]
+    gmail_ad.marketing_image_headline.value = 'Travel'
+    gmail_ad.marketing_image_description.value = 'Take to the skies!'
+
+    final_url = ad_group_ad.ad.final_urls.add()
+    final_url.value = 'http://www.example.com'
+    ad_group_ad.ad.name.value = 'Gmail Ad #{}'.format(str(uuid4())[:15])
```

Comment:
why [:15]? It cuts in the middle of time_hi_and_version. Should be either [:13] or [:18]. See https://en.wikipedia.org/wiki/Universally_unique_identifier#Format

Reply:
I took this from the Ruby example, I believe is just limits the length of the random string to 15 chars. This is only used to randomize the ad name so there aren't collisions. Since we use the entire uuid in other examples I'll take it out.